File: server/app/schemas/therapist.py
from pydantic import BaseModel, Field, field_validator
from typing import Optional


class DefectItem(BaseModel):
    defect_id: str
    code: str
    name: str
    category: str


class NotificationOut(BaseModel):
    id: str
    notification_type: str
    message: str
    is_read: bool
    created_at: str
    patient_id: Optional[str]
    attempt_id: Optional[str]


class PatientListItem(BaseModel):
    patient_id: str
    full_name: str
    email: str
    status: str
    date_of_birth: str
    gender: Optional[str]
    pre_assigned_defect_ids: Optional[dict]
    created_at: str

    model_config = {"from_attributes": True}


class ApprovePatientRequest(BaseModel):
    defect_ids: list[str] = Field(min_length=1)
    primary_diagnosis: Optional[str] = Field(default=None, max_length=200)
    clinical_notes: Optional[str] = Field(default=None, max_length=2000)

    @field_validator("defect_ids")
    @classmethod
    def validate_defect_ids(cls, value: list[str]) -> list[str]:
        cleaned: list[str] = []
        seen: set[str] = set()
        for defect_id in value:
            normalized = defect_id.strip()
            if not normalized:
                raise ValueError("defect_ids cannot contain blank values")
            if normalized not in seen:
                seen.add(normalized)
                cleaned.append(normalized)
        if not cleaned:
            raise ValueError("At least one defect must be selected")
        return cleaned

    @field_validator("primary_diagnosis", "clinical_notes", mode="before")
    @classmethod
    def normalize_optional_text(cls, value: Optional[str]) -> Optional[str]:
        if value is None:
            return None
        if isinstance(value, str):
            normalized = value.strip()
            return normalized or None
        return value


class DashboardResponse(BaseModel):
    total_patients: int
    approved_patients: int
    pending_patients: int
    patients_without_baseline: int = 0
    patients_without_approved_plan: int = 0
    plans_pending_approval: int = 0
    unread_notification_count: int = 0


class TherapistProfileResponse(BaseModel):
    therapist_id: str
    full_name: str
    email: str
    therapist_code: str
    license_number: Optional[str]
    specialization: Optional[str]
    years_of_experience: Optional[int]

File: client/types/therapist.ts
export interface Defect {
  defect_id: string;
  code: string;
  name: string;
  category: string;
}

export interface Task {
  task_id: string;
  name: string;
  type: string;
  task_mode: string;
  description: string | null;
}

export interface TherapistProfile {
  therapist_id: string;
  full_name: string;
  email: string;
  therapist_code: string;
  license_number: string | null;
  specialization: string | null;
  years_of_experience: number | null;
}

export interface TherapistDashboard {
  total_patients: number;
  approved_patients: number;
  pending_patients: number;
  patients_without_baseline: number;
  patients_without_approved_plan: number;
  plans_pending_approval: number;
  unread_notification_count: number;
}

export interface ApproveRequest {
  defect_ids: string[];
  primary_diagnosis?: string;
  clinical_notes?: string;
}

export interface Notification {
  id: string;
  notification_type: 'patient_registered' | 'review_flagged' | string;
  message: string;
  is_read: boolean;
  created_at: string;
  patient_id: string | null;
  attempt_id: string | null;
}

export interface AdaptationStep {
  from_level: string;
  to_level: string;
  attempts_used: number;
  reason: string;
  final_score: number;
}

export interface AdaptationEvent {
  session_id: string;
  session_date: string;
  task_id: string;
  task_name: string;
  adaptation_count: number;
  escalated: boolean;
  adaptation_history: AdaptationStep[];
  adaptation_report: Record<string, unknown> | null;
  linked_plan: RegeneratedPlan | null;
}

export interface RegeneratedAssignment {
  assignment_id: string;
  task_id: string;
  task_name: string;
  initial_level_name: string | null;
  day_index: number | null;
}

export interface RegeneratedPlan {
  plan_id: string;
  plan_name: string;
  status: "draft" | "approved" | "archived";
  created_at: string;
  regeneration_note: string | null;
  assignments: RegeneratedAssignment[];
}

export interface AdaptationActivity {
  adaptation_events: AdaptationEvent[];
  regenerated_plans: RegeneratedPlan[];
}
